### ISSUE TITLE
Reading vectors, read(io, Vector{T}, n)

### ISSUE THREAD

**Issue by oxinabox** [JuliaLang/julia#41865]
We have `read(io, T)` for all primate types.
That will read a value of type `T` from the `io`.

And we have `read(io, nb)` that will read a `Vector{UInt8}` of to length `nb` bytes from the stream.

But we don't have a one that would read a `Vector{T}` of length up the `n` elements of type `T`.
Even though we do have `write(io, ::Vector{T})`

**Comment by Moelf** [comment]
Is there anything better than this:
```julia
function read(io, Vector{T}, n) where T
    reinterpret(T, read(io, n))
end
```
which doesn't return "normal" vector.

We're pretty close to low-level here, maybe not much we can do without writing some C code:
https://github.com/JuliaLang/julia/blob/b4588945351019c89477192d1daa6e69304c51e9/base/iostream.jl#L487

**Comment by oxinabox** [comment]
Probably not, (but it is not about faster, it is about having a nice user interface.)

BTW I think, it needs to be, but same idea.
```
function read(io, Vector{T}, n) where T
    reinterpret(T, read(io, n*sizeof(T)))
end
```

**Comment by StefanKarpinski** [comment]
Also: avoiding reinterpret is generally good.

**Comment by simeonschaub** [comment]
Yes, I think `read!(io, Vector{T}(undef, n))` should be more efficient and reliable than using `reinterpret` here.

**Comment by Moelf** [comment]
>read!(io, Vector{T}(undef, n))

user knows "the rest will be a Vector{Float32}", but doesn't know how long is that

**Comment by mcabbott** [comment]
Shouldn't the spelling be `read(io, T, n)`? Like `read(io, n)` this would return a Vector, but specifying `Vector{T}` seems odd if you couldn't specify anything else.

**Comment by oxinabox** [comment]
I think `read(io, Vector{T}, n=typemax(Int))` so by default to read all.
so that the 2 arg version  doesn't clash with `read(io, T)`
